src/components/dom.js

```diff
@@ -98,4 +159,126 @@ export default class Dom {
 
     }
 
+    /**
+     * Checks target if it is native input
+     * @param {Element|*} target - HTML element or string
+     * @return {boolean}
+     */
+    static isNativeInput(target) {
+
+        let nativeInputs = [
+            'INPUT',
+            'TEXTAREA'
+        ];
+
+        return target ? nativeInputs.indexOf(target.tagName) !== -1 : false;
+
+    }
+
+    /**
+     * Checks node if it is empty
+     * It must be node without childNodes
+     * @param {Node} node
+     *
+     * @return {Boolean} true if it is empty
+     */
+    static checkNodeEmpty(node) {
+
+        let nodeText;
+
+        if ( this.isElement(node) && this.isNativeInput(node) ) {
+
+            nodeText = node.value;
+
+            if ( nodeText.trim() ) {
+
+                return false;
+
+            }
+
+
+        } else {
+
+            nodeText = node.textContent.replace('\u200B', '');
+
+            if ( nodeText.trim() ) {
+
+                return false;
+
+            }
+
+        }
+
+        return true;
+
+    }
+
+    /**
+     * checks node if it is doesn't have child node
+     * @param {Node} node
+     * @return {*|boolean}
+     */
+    static isLeaf(node) {
+
+        if (!node) {
+
+            return false;
+
+        }
+
+        return node.childNodes.length === 0;
+
+    }
+
+    /**
+     * breadth-first search
+     *
+     * Pushes to stack all DOM leafs and checks for emptiness
+     * @param {Node} node
+     * @return {boolean}
+     */
+    static isEmpty(node) {
+
+        let treeWalker = [],
+            stack = [];
+
+        if (!node) {
+
+            return false;
```

Comment:
`undefined` is not empty Node? Maybe you should throw an Error there? or add a `console.assert`